perllib/FixMyStreet/App/Controller/Admin/AreaStats.pm

```diff
@@ -0,0 +1,208 @@
+package FixMyStreet::App::Controller::Admin::AreaStats;
+use Moose;
+use namespace::autoclean;
+use List::Util qw(sum);
+
+BEGIN { extends 'Catalyst::Controller'; }
+
+sub begin : Private {
+    my ( $self, $c ) = @_;
+
+    $c->forward('/admin/begin');
+}
+
+sub index : Path : Args(0) {
+    my ( $self, $c ) = @_;
+
+    my $user = $c->user;
+
+    if ($user->is_superuser) {
+        $c->forward('/admin/fetch_all_bodies');
+    } elsif ( $user->from_body ) {
+        $c->forward('load_user_body', [ $user->from_body->id ]);
+        $c->stash->{body_id} = $user->from_body->id;
+        if ($user->area_id) {
+            $c->stash->{area_id} = $user->area_id;
+            $c->visit( 'stats' );
+        } else {
+            # visit body_stats so we load the list of child areas
+            $c->visit( 'body_stats', [ $user->from_body->id], [] );
+        }
+    } else {
+        $c->detach( '/page_error_404_not_found' );
+    }
+}
+
+sub check_user : Private {
+    my ( $self, $c, $body_id, $area_id ) = @_;
+
+    my $user = $c->user;
+
+    if ($user->is_superuser) {
+        return;
+    }
+
+    if ($body_id and $user->from_body->id eq $body_id) {
+        if (not $user->area_id) {
+            return;
+        } elsif ($area_id and $user->area_id eq $area_id) {
+            return;
+        }
+    }
+
+    $c->detach( '/page_error_404_not_found' );
+}
+
+sub body_base : Chained('/') : PathPart('admin/areastats') : CaptureArgs(1) {
+    my ($self, $c, $body_id) = @_;
+    $c->stash->{body_id} = $body_id;
+    $c->forward('/admin/lookup_body');
+    $c->stash->{areas} = mySociety::MaPit::call('area/children', [ $body_id ] );
+}
+
+sub body_stats : Chained('body_base') : PathPart('') : Args(0) {
+    my ($self, $c) = @_;
+    my $user = $c->user;
+
+    $c->forward('check_user', [$c->stash->{body_id}]);
+    if ($c->get_param('area')) {
+        $c->stash->{area_id} = $c->get_param('area');
+    } else {
+        $c->stash->{area_id} = $c->stash->{body}->id;
+    }
+    $c->forward('stats');
+}
+
+sub ward_stats : Chained('body_base') :PathPart('') : Args(1) {
+    my ($self, $c, $area_id) = @_;
+    my $user = $c->user;
+
+    $c->forward('check_user', [$c->stash->{body_id}, $area_id]);
+
+    $c->stash->{area_id} = $area_id;
+    $c->forward('stats');
+}
+
+sub stats : Private {
+    my ($self, $c) = @_;
+
+    my $date = DateTime->now->subtract(days => 30);
+    my $area_id = $c->stash->{area_id};
+    my $area = mySociety::MaPit::call('area', $area_id );
+    my $user = $c->user;
+
+    $c->forward('/admin/fetch_contacts');
+
+    $c->stash->{template} = 'admin/areastats/area.html';
+    if ($area->{name}) {
```

Comment:
You could detach the 404 here and not indent the rest of the function :) (But I think this will be changing anyway.)

Reply:
✅